apis/python/src/tiledbsoma/tiledb_object.py

```diff
@@ -113,15 +125,100 @@ def exists(self) -> bool:
         # before a third, successful HTTP request for group-open.  Instead, we directly attempt the
         # group-open request, checking for an exception.
         try:
-            return (
-                self._metadata.get(util.SOMA_OBJECT_TYPE_METADATA_KEY) == self.soma_type
-            )
+            with self._maybe_open():
+                return (
+                    self._metadata.get(util.SOMA_OBJECT_TYPE_METADATA_KEY)
+                    == self.soma_type
+                )
         except tiledb.cc.TileDBError:
             return False
 
-    @abstractmethod
-    def _tiledb_open(self, mode: str = "r") -> Union[tiledb.Array, tiledb.Group]:
-        """Open the underlying TileDB array or Group"""
+    def open(self: TTileDBObject, mode: str = "r") -> TTileDBObject:
+        """
+        Open the object for a series of read or write operations. This is optional, but makes a
+        series of small operations more efficient by allowing reuse of handles and metadata. If
+        opened, the object should later be closed to release such resources. This can be automated
+        with a context manager, e.g.:
+
+          with tiledbsoma.SparseNDArray(uri=...).open() as X:
+              data = X.read_table(...)
+              data = X.read_table(...)
+
+        If the object is not opened, then any individual read or write operation effectively opens
+        and closes it transiently for the operation.
+        """
+        if mode not in ("r", "w"):
+            raise ValueError(
+                self.__class__.__name__ + " open mode must be one of 'r', 'w'"
+            )
+        if self._open_mode:
+            raise RuntimeError(self.__class__.__name__ + " is already open")
+        self._open_mode = mode
+        return self
+
+    def close(self) -> None:
+        """
+        Release any resources held while the object is open. Closing an already-closed object is a
+        no-op.
+        """
+        if self._open_mode:
+            self._open_mode = ""
+            self._close_stack.close()
+
+    def __enter__(self: TTileDBObject) -> TTileDBObject:
+        assert (
+            self._open_mode
+        ), f"use {self.__class__.__name__}.open() as context manager"
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        if self._open_mode:
```

Comment:
given that close is idempotent, I'd just call it without the conditional.  Style choice only, so totally up to you.